src/battle_ai_main.c

```diff
@@ -2645,6 +2645,13 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
             if (IsDynamaxed(battlerDef))
                 ADJUST_SCORE(-10);
             break;
+        case EFFECT_UPPER_HAND:
+            if (predictedMove != MOVE_NONE)
+            {
+                if (IS_MOVE_STATUS(predictedMove) || AI_WhoStrikesFirst(battlerAtk, battlerDef, move) == AI_IS_SLOWER || gMovesInfo[predictedMove].priority < 1 || gMovesInfo[predictedMove].priority > 3) // Opponent going first or not using priority move
```

Comment:
You want to use an actual priority calc instead of using `gMovesInfo[predictedMove].priority` because the function will handle things like Gale Wings.